File: app/src/main/java/org/schabi/newpipe/player/helper/LoadController.java
package org.schabi.newpipe.player.helper;

import com.google.android.exoplayer2.DefaultLoadControl;

public class LoadController extends DefaultLoadControl {

    public static final String TAG = "LoadController";
    private boolean preloadingEnabled = true;

    @Override
    public void onPrepared() {
        preloadingEnabled = true;
        super.onPrepared();
    }

    @Override
    public void onStopped() {
        preloadingEnabled = true;
        super.onStopped();
    }

    @Override
    public void onReleased() {
        preloadingEnabled = true;
        super.onReleased();
    }

    @Override
    public boolean shouldContinueLoading(final long playbackPositionUs,
                                         final long bufferedDurationUs,
                                         final float playbackSpeed) {
        if (!preloadingEnabled) {
            return false;
        }
        return super.shouldContinueLoading(
                playbackPositionUs, bufferedDurationUs, playbackSpeed);
    }

    public void disablePreloadingOfCurrentTrack() {
        preloadingEnabled = false;
    }
}

File: app/src/main/java/org/schabi/newpipe/player/helper/LoadController.kt
@file:OptIn(androidx.media3.common.util.UnstableApi::class)

package org.schabi.newpipe.player.helper

import androidx.media3.exoplayer.DefaultLoadControl

class LoadController : DefaultLoadControl() {

    private var preloadingEnabled = true

    override fun onPrepared() {
        preloadingEnabled = true
        super.onPrepared()
    }

    override fun onStopped() {
        preloadingEnabled = true
        super.onStopped()
    }

    override fun onReleased() {
        preloadingEnabled = true
        super.onReleased()
    }

    override fun shouldContinueLoading(
        playbackPositionUs: Long,
        bufferedDurationUs: Long,
        playbackSpeed: Float
    ): Boolean {
        if (!preloadingEnabled) {
            return false
        }
        return super.shouldContinueLoading(
            playbackPositionUs,
            bufferedDurationUs,
            playbackSpeed
        )
    }

    fun disablePreloadingOfCurrentTrack() {
        preloadingEnabled = false
    }

    companion object {
        const val TAG = "LoadController"
    }
}
